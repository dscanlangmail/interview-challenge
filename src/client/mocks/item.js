export const mockItem = {
  id: 1001,
  name: "Kale Caesar Pasta, Turmeric Satay Broccoli & Lemon Cashew Greens",
  dietaries: ["ve", "df", "gf", "n!"],
};
