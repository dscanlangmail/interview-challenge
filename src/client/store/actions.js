import { ActionType, createAction } from "typesafe-actions";

export const ActionTypes = {
  SELECT_ITEM: "SELECT_ITEM",
  REMOVE_ITEM: "REMOVE_ITEM",
};

export const selectItem = createAction(ActionTypes.SELECT_ITEM)();
export const removeItem = createAction(ActionTypes.REMOVE_ITEM)();
