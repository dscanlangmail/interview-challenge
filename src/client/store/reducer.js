import { getType } from "typesafe-actions";
import { ActionTypes } from "./actions";

const initialState = [];

export const reducer = (state = initialState, action) => {
  switch (action.type) {
    case ActionTypes.SELECT_ITEM: {
      return [...state, action.payload];
    }
    case ActionTypes.REMOVE_ITEM: {
      const newState = [...state];
      const index = newState.findIndex((item) => {
        return item.id === action.payload.id;
      });

      newState.splice(index, 1);

      return newState;
    }
    default:
      [];
  }
};
