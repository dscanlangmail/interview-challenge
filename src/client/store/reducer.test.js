import { reducer as foodReducer } from "./reducer";
import { ActionTypes } from "./actions";
import { mockItem } from "../mocks/item";

describe("reducer", () => {
  it("should return the initial state", () => {
    expect(foodReducer(undefined, {})).toMatchSnapshot();
  });

  it("should handle selecting a new item", () => {
    expect(
      foodReducer([], { type: ActionTypes.SELECT_ITEM, payload: mockItem })
    ).toMatchSnapshot();
  });
  it("should handle removing an item", () => {
    expect(
      foodReducer([mockItem], {
        type: ActionTypes.REMOVE_ITEM,
        payload: mockItem,
      })
    ).toMatchSnapshot();
  });
});
