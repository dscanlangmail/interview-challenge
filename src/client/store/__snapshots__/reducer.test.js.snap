// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`reducer should handle removing an item 1`] = `Array []`;

exports[`reducer should handle selecting a new item 1`] = `
Array [
  Object {
    "dietaries": Array [
      "ve",
      "df",
      "gf",
      "n!",
    ],
    "id": 1001,
    "name": "Kale Caesar Pasta, Turmeric Satay Broccoli & Lemon Cashew Greens",
  },
]
`;

exports[`reducer should return the initial state 1`] = `undefined`;
